File: packages/php-sdk/lib/BillinProvider.php
<?php
/**
 * BillinProvider
 * PHP version 7.4 | 8.0+
 *
 * @category Class
 * @package  FacturasBillin\SDK
 * @author   OpenAPI Generator team
 * @link     https://openapi-generator.tech
 */

/**
 * Billin Api
 *
 * Simple wrapper for the Billin API SDK.
 *
 * The version of the OpenAPI document: 1.0
 */

namespace FacturasBillin\SDK;

use GuzzleHttp\Client;
use GuzzleHttp\ClientInterface;
use FacturasBillin\SDK\Api\AuthApi;
use FacturasBillin\SDK\Api\InvoicesApi;
use FacturasBillin\SDK\Api\ExpensesApi;
use FacturasBillin\SDK\Api\ReceiptsApi;
use FacturasBillin\SDK\Api\ContactsApi;
use FacturasBillin\SDK\Api\ProductsApi;
use FacturasBillin\SDK\Api\PaymentsApi;
use FacturasBillin\SDK\Api\AccountingAccountsApi;
use FacturasBillin\SDK\Api\QuotesApi;

/**
 * BillinProvider Class Doc Comment
 *
 * Simple wrapper for the Billin API SDK.
 *
 * This provider offers a simplified interface to interact with the Billin API.
 * Users can authenticate and perform operations like creating invoices, expenses, receipts, etc.
 *
 * Example usage:
 * <code>
 * $provider = new BillinProvider('your-api-token');
 *
 * // Create an invoice
 * $invoice = new \FacturasBillin\SDK\Model\InvoiceCreateDto();
 * // ... configure invoice details
 * $response = $provider->getInvoicesApi()->createInvoice($invoice);
 * </code>
 *
 * @category Class
 * @package  FacturasBillin\SDK
 * @author   jamataran
 * @link     https://github.com/jamataran/facturas-billin-sdk
 */
class BillinProvider
{
    /**
     * @var Configuration
     */
    private $config;

    /**
     * @var ClientInterface
     */
    private $httpClient;

    /**
     * @var AuthApi
     */
    private $authApi;

    /**
     * @var InvoicesApi
     */
    private $invoicesApi;

    /**
     * @var ExpensesApi
     */
    private $expensesApi;

    /**
     * @var ReceiptsApi
     */
    private $receiptsApi;

    /**
     * @var ContactsApi
     */
    private $contactsApi;

    /**
     * @var ProductsApi
     */
    private $productsApi;

    /**
     * @var PaymentsApi
     */
    private $paymentsApi;

    /**
     * @var AccountingAccountsApi
     */
    private $accountingAccountsApi;

    /**
     * @var QuotesApi
     */
    private $quotesApi;

    /**
     * Default base URL for the Billin API
     */
    public const DEFAULT_BASE_URL = 'https://api.billin.net';

    /**
     * Creates a new BillinProvider with the specified API token.
     *
     * @param string               $apiToken   The API token for authentication (Bearer token)
     * @param string               $baseUrl    The base URL of the Billin API (optional)
     * @param ClientInterface|null $httpClient Custom Guzzle HTTP client (optional)
     */
    public function __construct(
        string $apiToken,
        string $baseUrl = self::DEFAULT_BASE_URL,
        ?ClientInterface $httpClient = null
    ) {
        $this->config = new Configuration();
        $this->config->setHost($baseUrl);
        $this->config->setAccessToken($apiToken);

        $this->httpClient = $httpClient ?? new Client();

        // Initialize API clients
        $this->initializeApis();
    }

    /**
     * Initialize all API clients
     *
     * @return void
     */
    private function initializeApis(): void
    {
        $this->authApi = new AuthApi($this->httpClient, $this->config);
        $this->invoicesApi = new InvoicesApi($this->httpClient, $this->config);
        $this->expensesApi = new ExpensesApi($this->httpClient, $this->config);
        $this->receiptsApi = new ReceiptsApi($this->httpClient, $this->config);
        $this->contactsApi = new ContactsApi($this->httpClient, $this->config);
        $this->productsApi = new ProductsApi($this->httpClient, $this->config);
        $this->paymentsApi = new PaymentsApi($this->httpClient, $this->config);
        $this->accountingAccountsApi = new AccountingAccountsApi($this->httpClient, $this->config);
        $this->quotesApi = new QuotesApi($this->httpClient, $this->config);
    }

    /**
     * Gets the configuration object for advanced usage.
     *
     * @return Configuration
     */
    public function getConfig(): Configuration
    {
        return $this->config;
    }

    /**
     * Gets the underlying HTTP client.
     *
     * @return ClientInterface
     */
    public function getHttpClient(): ClientInterface
    {
        return $this->httpClient;
    }

    /**
     * Gets the Auth API client for authentication operations.
     *
     * @return AuthApi
     */
    public function getAuthApi(): AuthApi
    {
        return $this->authApi;
    }

    /**
     * Gets the Invoices API client for invoice operations.
     *
     * @return InvoicesApi
     */
    public function getInvoicesApi(): InvoicesApi
    {
        return $this->invoicesApi;
    }

    /**
     * Gets the Expenses API client for expense operations.
     *
     * @return ExpensesApi
     */
    public function getExpensesApi(): ExpensesApi
    {
        return $this->expensesApi;
    }

    /**
     * Gets the Receipts API client for receipt/sales receipt operations.
     *
     * @return ReceiptsApi
     */
    public function getReceiptsApi(): ReceiptsApi
    {
        return $this->receiptsApi;
    }

    /**
     * Gets the Contacts API client for contact operations.
     *
     * @return ContactsApi
     */
    public function getContactsApi(): ContactsApi
    {
        return $this->contactsApi;
    }

    /**
     * Gets the Products API client for product operations.
     *
     * @return ProductsApi
     */
    public function getProductsApi(): ProductsApi
    {
        return $this->productsApi;
    }

    /**
     * Gets the Payments API client for payment operations.
     *
     * @return PaymentsApi
     */
    public function getPaymentsApi(): PaymentsApi
    {
        return $this->paymentsApi;
    }

    /**
     * Gets the Accounting Accounts API client for accounting account operations.
     *
     * @return AccountingAccountsApi
     */
    public function getAccountingAccountsApi(): AccountingAccountsApi
    {
        return $this->accountingAccountsApi;
    }

    /**
     * Gets the Quotes API client for quote operations.
     *
     * @return QuotesApi
     */
    public function getQuotesApi(): QuotesApi
    {
        return $this->quotesApi;
    }

    /**
     * Sets the base path for the API client.
     *
     * @param string $basePath The base path
     * @return $this
     */
    public function setBasePath(string $basePath): self
    {
        $this->config->setHost($basePath);
        $this->initializeApis();
        return $this;
    }

    /**
     * Sets the bearer token for authentication.
     *
     * @param string $token The bearer token
     * @return $this
     */
    public function setBearerToken(string $token): self
    {
        $this->config->setAccessToken($token);
        $this->initializeApis();
        return $this;
    }

    /**
     * Creates a BillinProvider by authenticating with client credentials.
     *
     * This is a factory method that handles the OAuth2 client_credentials flow.
     *
     * @param string               $clientId     The client ID
     * @param string               $clientSecret The client secret
     * @param string               $baseUrl      The base URL of the Billin API (optional)
     * @param ClientInterface|null $httpClient   Custom Guzzle HTTP client (optional)
     * @return BillinProvider
     * @throws ApiException If authentication fails
     */
    public static function withClientCredentials(
        string $clientId,
        string $clientSecret,
        string $baseUrl = self::DEFAULT_BASE_URL,
        ?ClientInterface $httpClient = null
    ): BillinProvider {
        $config = new Configuration();
        $config->setHost($baseUrl);

        $client = $httpClient ?? new Client();
        $authApi = new AuthApi($client, $config);

        // Create token request
        $tokenRequest = new \FacturasBillin\SDK\Model\RequestTokenUserApiCredentialDto();
        $tokenRequest->setClientId($clientId);
        $tokenRequest->setClientSecret($clientSecret);
        $tokenRequest->setGrantType(\FacturasBillin\SDK\Model\RequestTokenUserApiCredentialDto::GRANT_TYPE_CLIENT_CREDENTIALS);

        // Get access token
        $tokenResponse = $authApi->authControllerGetTokenV1($tokenRequest);
        $accessToken = $tokenResponse->getAccessToken();

        return new BillinProvider($accessToken, $baseUrl, $client);
    }

    /**
     * Creates a BillinProvider by authenticating with username and password.
     *
     * This is a factory method that handles the OAuth2 password flow.
     *
     * @param string               $clientId     The client ID
     * @param string               $clientSecret The client secret
     * @param string               $username     The username
     * @param string               $password     The password
     * @param string               $baseUrl      The base URL of the Billin API (optional)
     * @param ClientInterface|null $httpClient   Custom Guzzle HTTP client (optional)
     * @return BillinProvider
     * @throws ApiException If authentication fails
     */
    public static function withPassword(
        string $clientId,
        string $clientSecret,
        string $username,
        string $password,
        string $baseUrl = self::DEFAULT_BASE_URL,
        ?ClientInterface $httpClient = null
    ): BillinProvider {
        $config = new Configuration();
        $config->setHost($baseUrl);

        $client = $httpClient ?? new Client();
        $authApi = new AuthApi($client, $config);

        // Create token request
        $tokenRequest = new \FacturasBillin\SDK\Model\RequestTokenUserApiCredentialDto();
        $tokenRequest->setClientId($clientId);
        $tokenRequest->setClientSecret($clientSecret);
        $tokenRequest->setUsername($username);
        $tokenRequest->setPassword($password);
        $tokenRequest->setGrantType(\FacturasBillin\SDK\Model\RequestTokenUserApiCredentialDto::GRANT_TYPE_PASSWORD);

        // Get access token
        $tokenResponse = $authApi->authControllerGetTokenV1($tokenRequest);
        $accessToken = $tokenResponse->getAccessToken();

        return new BillinProvider($accessToken, $baseUrl, $client);
    }
}

File: packages/java-sdk/src/main/java/net/facturasbillin/sdk/BillinProvider.java
package net.facturasbillin.sdk;

import net.facturasbillin.sdk.api.AuthApi;
import net.facturasbillin.sdk.api.InvoicesApi;
import net.facturasbillin.sdk.api.ExpensesApi;
import net.facturasbillin.sdk.api.ReceiptsApi;
import net.facturasbillin.sdk.api.ContactsApi;
import net.facturasbillin.sdk.api.ProductsApi;
import net.facturasbillin.sdk.api.PaymentsApi;
import net.facturasbillin.sdk.api.AccountingAccountsApi;
import net.facturasbillin.sdk.api.QuotesApi;
import net.facturasbillin.sdk.model.*;
import net.facturasbillin.sdk.auth.HttpBearerAuth;
import okhttp3.OkHttpClient;

/**
 * Simple wrapper for the Billin API SDK.
 *
 * This provider offers a simplified interface to interact with the Billin API.
 * Users can authenticate and perform operations like creating invoices, expenses, receipts, etc.
 *
 * Example usage:
 * <pre>{@code
 * BillinProvider provider = new BillinProvider("your-api-token");
 *
 * // Create an invoice
 * InvoiceCreateDto invoice = new InvoiceCreateDto();
 * // ... configure invoice details
 * GetInvoice200Response response = provider.getInvoicesApi().createInvoice(invoice);
 * InvoiceDocumentAnswerDto createdInvoice = response.getInvoice();
 * }</pre>
 */
public class BillinProvider {

    private final ApiClient apiClient;
    private final InvoicesApi invoicesApi;
    private final ExpensesApi expensesApi;
    private final ReceiptsApi receiptsApi;
    private final ContactsApi contactsApi;
    private final ProductsApi productsApi;
    private final PaymentsApi paymentsApi;
    private final AccountingAccountsApi accountingAccountsApi;
    private final QuotesApi quotesApi;
    private String bearerToken;

    /**
     * Creates a new BillinProvider with the specified API token.
     *
     * @param apiToken the API token for authentication
     */
    public BillinProvider(String apiToken) {
        this(apiToken, "https://api.billin.net");
    }

    /**
     * Creates a new BillinProvider with the specified API token and base URL.
     *
     * @param apiToken the API token for authentication
     * @param baseUrl  the base URL of the Billin API
     */
    public BillinProvider(String apiToken, String baseUrl) {
        this.bearerToken = apiToken;
        this.apiClient = new ApiClient();
        this.apiClient.setBasePath(baseUrl);

        // Configure authentication
        HttpBearerAuth bearerAuth = (HttpBearerAuth) apiClient.getAuthentication("bearer");
        if (bearerAuth != null) {
            bearerAuth.setBearerToken(apiToken);
        }

        // Initialize API clients
        this.invoicesApi = new InvoicesApi(apiClient);
        this.expensesApi = new ExpensesApi(apiClient);
        this.receiptsApi = new ReceiptsApi(apiClient);
        this.contactsApi = new ContactsApi(apiClient);
        this.productsApi = new ProductsApi(apiClient);
        this.paymentsApi = new PaymentsApi(apiClient);
        this.accountingAccountsApi = new AccountingAccountsApi(apiClient);
        this.quotesApi = new QuotesApi(apiClient);
    }


    /**
     * Gets the underlying API client for advanced usage.
     *
     * @return the API client
     */
    public ApiClient getApiClient() {
        return apiClient;
    }

    /**
     * Gets the invoices API client for direct access.
     *
     * @return the invoices API client
     */
    public InvoicesApi getInvoicesApi() {
        return invoicesApi;
    }

    /**
     * Gets the expenses API client for direct access.
     *
     * @return the expenses API client
     */
    public ExpensesApi getExpensesApi() {
        return expensesApi;
    }

    /**
     * Gets the receipts API client for direct access.
     *
     * @return the receipts API client
     */
    public ReceiptsApi getReceiptsApi() {
        return receiptsApi;
    }

    /**
     * Gets the contacts API client for direct access.
     *
     * @return the contacts API client
     */
    public ContactsApi getContactsApi() {
        return contactsApi;
    }

    /**
     * Gets the products API client for direct access.
     *
     * @return the products API client
     */
    public ProductsApi getProductsApi() {
        return productsApi;
    }

    /**
     * Gets the payments API client for direct access.
     *
     * @return the payments API client
     */
    public PaymentsApi getPaymentsApi() {
        return paymentsApi;
    }

    /**
     * Gets the accounting accounts API client for direct access.
     *
     * @return the accounting accounts API client
     */
    public AccountingAccountsApi getAccountingAccountsApi() {
        return accountingAccountsApi;
    }

    /**
     * Gets the quotes API client for direct access.
     *
     * @return the quotes API client
     */
    public QuotesApi getQuotesApi() {
        return quotesApi;
    }

    /**
     * Sets the base path for the API client.
     *
     * @param basePath the base path
     */
    public void setBasePath(String basePath) {
        apiClient.setBasePath(basePath);
    }

    /**
     * Sets the bearer token for authentication.
     *
     * @param token the bearer token
     */
    public void setBearerToken(String token) {
        this.bearerToken = token;
        HttpBearerAuth bearerAuth = (HttpBearerAuth) apiClient.getAuthentication("bearer");
        if (bearerAuth != null) {
            bearerAuth.setBearerToken(token);
        }
    }

    /**
     * Gets the underlying OkHttp client for advanced customization.
     *
     * @return the OkHttp client
     */
    public OkHttpClient getHttpClient() {
        return apiClient.getHttpClient();
    }
}

